compiler/luci/service/src/CircleShapeInferenceHelper.cpp

```diff
@@ -16,6 +16,104 @@
 
 #include "CircleShapeInferenceHelper.h"
 
+#include <oops/InternalExn.h>
+
+using namespace luci::sinf;
+
+namespace
+{
+
+/**
+ * @brief  Expand shape x and y to same rank by align right and filling with 1
+ */
+void expand_rank(loco::TensorShape &x, loco::TensorShape &y)
+{
+  auto x_rank = x.rank();
+  auto y_rank = y.rank();
+
+  if (x_rank == y_rank)
+    return;
+
+  TensorShapeExpander x_exp(x);
+  TensorShapeExpander y_exp(y);
+
+  auto xy_rank = std::max(x_rank, y_rank);
+
+  x = x_rank > y_rank ? x : x_exp.to(xy_rank);
+  y = y_rank > x_rank ? y : y_exp.to(xy_rank);
+}
+
+/**
+ * @brief  Return shape of expanded dimension of input x and y having same rank
+ */
+loco::TensorShape expand_dimension(const loco::TensorShape &x, const loco::TensorShape &y)
+{
+  assert(x.rank() == y.rank()); // FIX_CALLER_UNLESS
+
+  auto rank = x.rank();
+
+  loco::TensorShape output_shape;
+
+  output_shape.rank(rank);
+
+  // Shape inference rule (commutative)
+  // Values of x, y, and Result are categorized as 1, X, ? where
+  // X is a positive integer greater than 1 and ? is unknown.
+  //     x | y | Result
+  // c1. 1 | X | X
+  // c2. 1 | ? | ?
+  // c3. X | X | X
+  // c4. X | ? | X
```

Comment:
```suggestion
  // c4. X | ? | ?
```

Is there a reason for c4?
I first thought that `Result` might be `?` in this case.

Reply:
`?` should be one of the below values in runtime.
- 1 -> In this case, output is X
- non-1 (!= X) -> In this case, exception has to be thrown
- non-1 (== X) -> In this case, output is X
So, the output has to be X if valid inputs are given.

It would be not simple to review the logic thoroughly. Fortunately, we have a good reference in TF (https://github.com/tensorflow/tensorflow/blob/v2.17.0/third_party/xla/xla/service/shape_inference.cc#L1037-L1070). I hope the code will help your review (and implementation).

Reply:
> So, the output has to be X if valid inputs are given.

Aha, definitely! I got it.